src/agentscope/agents/react_agent.py

```diff
@@ -96,230 +59,185 @@ def __init__(
-            self.speak(
-                Msg(self.name, json.dumps(res, indent=4), role="assistant"),
-            )
-
-            if self.memory:
-                self.memory.add(msg_thought)
+            # Step 2: Acting
+            if self.verbose:
+                self.speak(f" ITER {_+1}, STEP 2: ACTING ".center(70, "#"))
 
-            # Skip the next steps if no need to call tools
-            if len(res.get("function", [])) == 0:
-                return msg_thought
-
-            # Step 2: Action
-
-            self.speak(f" ITER {_+1}, STEP 2: ACTION ".center(70, "#"))
-
-            # Execute functions
-            # TODO: check the provided arguments and re-correct them if needed
-            execute_results = []
-            for i, func in enumerate(res["function"]):
-                # Execute the function
-                func_res = self.execute_func(i, func)
-                execute_results.append(func_res)
-
-            # Prepare prompt for execution results
-            execute_results_prompt = "\n".join(
-                [
-                    FUNCTION_RESULT_PROMPT.format_map(res)
-                    for res in execute_results
-                ],
-            )
-            # Add title
-            execute_results_prompt = (
-                FUNCTION_RESULT_TITLE_PROMPT + execute_results_prompt
-            )
+            # Parse, check and execute the tool functions in service toolkit
+            try:
+                execute_results = self.service_toolkit.parse_and_call_func(
+                    res.parsed["function"],
+                )
 
-            # Note: Observing the execution results and generate response are
-            # finished in the next loop. We just put the execution results
-            # into memory, and wait for the next loop to generate response.
+                # Note: Observing the execution results and generate response
+                # are finished in the next reasoning step. We just put the
+                # execution results into memory, and wait for the next loop
+                # to generate response.
 
-            # Record execution results into memory as a message from the system
-            msg_res = Msg(
-                name="system",
-                content=execute_results_prompt,
-                role="system",
-            )
-            self.speak(msg_res)
-            if self.memory:
+                # Record execution results into memory as system message
+                msg_res = Msg("system", execute_results, "system")
+                self.speak(msg_res)
                 self.memory.add(msg_res)
 
+            except FunctionCallError as e:
+                # Catch the function calling error that can be handled by
+                # the model
+                error_msg = Msg("system", str(e), "system")
+                self.speak(error_msg)
+                self.memory.add(error_msg)
+
         return Msg(
```

Comment:
Maybe we need to consider what is the more meaningful return of ReAct agent at this time. For the current version, other cooperative agents seems can not learn whether the execution is successful or not and what is the consequence.

Reply:
Why should other agents know about the tools calling and execution status within react agent?